common/darkside.go

```diff
@@ -5,49 +5,176 @@ import (
 }
 
-var state *DarksideZcashdState = nil
+var state darksideState
 
-func DarkSideRawRequest(method string, params []json.RawMessage) (json.RawMessage, error) {
+func DarksideIsEnabled() bool {
+	return state.chainName != ""
+}
 
-	if state == nil {
-		state = &DarksideZcashdState{
-			start_height:          1000,
-			sapling_activation:    1000,
-			branch_id:             "2bb40e60", // Blossom
-			chain_name:            "darkside",
-			blocks:                make([]string, 0),
-			incoming_transactions: make([][]byte, 0),
-			server_start:          time.Now(),
-		}
+func DarksideInit() {
+	state = darksideState{
+		startHeight:          -1,
+		saplingActivation:    -1,
+		branchID:             "2bb40e60", // Blossom
+		chainName:            "darkside",
+		blocks:               make([][]byte, 0),
+		incomingTransactions: make([][]byte, 0),
+		serverStart:          time.Now(),
+	}
+	RawRequest = darksideRawRequest
+	f := "testdata/darkside/init-blocks"
+	testBlocks, err := os.Open(f)
+	if err != nil {
+		Log.Warn("Error opening default darksidewalletd blocks file", f)
+		return
+	}
+	if err = readBlocks(testBlocks); err != nil {
+		Log.Warn("Error loading default darksidewalletd blocks")
+	}
+}
 
-		testBlocks, err := os.Open("./testdata/default-darkside-blocks")
+// DarksideAddBlock adds a single block to the blocks list.
+func DarksideAddBlock(blockHex string) error {
+	if blockHex == "404: Not Found" {
+		// special case error (http resource not found, bad pathname)
+		return errors.New(blockHex)
+	}
+	blockData, err := hex.DecodeString(blockHex)
+	if err != nil {
+		return err
+	}
+	block := parser.NewBlock()
+	rest, err := block.ParseFromSlice(blockData)
+	if err != nil {
+		return err
+	}
+	if len(rest) != 0 {
+		return errors.New("block serialization is too long")
+	}
+	blockHeight := block.GetHeight()
+	// first block, add to existing blocks slice if possible
+	if blockHeight > state.startHeight+len(state.blocks) {
+		// The new block can't contiguously extend the existing
+		// range, so we have to drop the existing range.
+		state.blocks = state.blocks[:0]
+	} else if blockHeight < state.startHeight {
+		// This block will replace the entire existing range.
+		state.blocks = state.blocks[:0]
+	} else {
+		// Drop the block that will be overwritten, and its children.
+		state.blocks = state.blocks[:blockHeight-state.startHeight]
+	}
+	if len(state.blocks) == 0 {
+		state.startHeight = blockHeight
+	} else {
+		// Set this block's prevhash.
+		prevblock := parser.NewBlock()
+		rest, err := prevblock.ParseFromSlice(state.blocks[len(state.blocks)-1])
 		if err != nil {
-			Log.Fatal("Error loading default darksidewalletd blocks")
+			return err
 		}
-		scan := bufio.NewScanner(testBlocks)
-		for scan.Scan() { // each line (block)
-			block := scan.Bytes()
-			state.blocks = append(state.blocks, string(block))
+		if len(rest) != 0 {
+			return errors.New("block is too long")
 		}
+		copy(blockData[4:4+32], prevblock.GetEncodableHash())
```

Comment:
I'm thinking that we shouldn't automatically patch up the `prevHash` here. By not doing so, testers retain full control over the blocks' prevHashes (aside from constraints on them set by how lightwalletd handles reorgs). That might be useful for testing lightwalletd or wallets for very-misbehaving zcashd behavior. Also by not doing it here, it would make it obvious when our dataset blocks checked into the other repo have invalid prevHashes.

Reply:
I see what you mean, but somehow I'm not convinced. Yes, it's good to throw bad input at the code under test to see how it reacts, which is the whole idea of fuzzing. But this isn't bad input. This can happen in real life with a zcashd that isn't misbehaving. The lightwalletd can only interpret an incorrect prevhash as a reorg, which is already being tested in various ways. The difference would be that this would be a _persistent_ incorrect prevhash, but that's not qualitatively different -- it's theoretically _possible_ for 100 one-block reorgs to occur in a row. 

After 100 consecutive reorgs (from lightwalletd's point of view), which occurs automatically by zcashd insisting on presenting a block with the wrong prevhash long enough, lightwalletd will crash. (I'm not sure it should, by the way, but that's a separate topic.) Do we want tests that cause lightwalletd to crash? There are no such tests so far, but if we do want such a test, how will we detect lightwalletd crashed, and crashed for the right reason? Open and inspect the log file? Also, how long do we wait? The lightwalletd sleeps a few seconds between attempts to resolve a reorg, which makes the timing to reach 100 reorgs somewhat random. (Such a test will take a few hundred seconds, which is undesirable.)

I added this prevhash patching to make it more convenient for the tests so they wouldn't have to always submit blocks from the beginning. That requirement seems inefficient and brittle to me. I'd like to move to the `Generate()` gRPC that would create a single block at a time. (Speaking of which, the zcash Python rpc-tests have a `generate` test RPC, which inspired my suggestion for `Generate()`, and that doesn't provide a way to create an incorrect prevhash. Has that caused a testing deficiency? That's an admittedly weak point, but thought I'd mention it.)

Reply:
I agree about how little use it would be for testing. The only case we're missing out on is a case where zcashd is persistently presenting data with a broken `prevHash` which should never happen.

Another way to say what I meant is: If this bit of code actually fixes up the `prevHash`, it means the set of blocks we loaded into darksidewalletd has a broken `prevHash` somewhere. Whoever made that dataset probably wants to find out quickly instead of having their mistake hidden from them (since it's a sign there could be *other* problems with the dataset). For `Generate()`, its implementation can bare the burden of using the correct `prevHash` before passing the block to this function.

Maybe a compromise is to emit log messages if the prevHash was not actually correct?

By the way, you might be interested in the background that leads me to have the opposite intuition: http://langsec.org/papers/postel-patch.pdf. tl;dr: There are two competing design philosophies, one is "handle unexpected errors gracefully" and the other is "crash and make a lot of noise when anything unexpected happens". It's harder to get things up and running by following the latter principle, but it has several advantages: code ends up being simpler, it's easier to test because there aren't as many edge cases, and bugs in Component A aren't hidden by Component B's graceful handling of Component A's misbehavior.